### PR TITLE
Add hashing and cryptography libraries

### PR DESCRIPTION
**Pull request by josesimoes**
## Description
- Add both assembly declarations to CMake.
- Add implementation for STM32 with ChibiOS.
- Add implementation for ESP32.
- Add implementation for Silabs Giant Gecko in Azure RTOS.
- Update build options in several platforms/targets.
- Rework CRC32 calculator in ChibiOS nano HAL implementation.
- Regular CRC32 is now setting the config params all the time.
- Rework CRC32 calculator in Silabs Giant Gecko platform implementation.
- Move adding Mbed TLS to a binutil function.
- Add call from ChibiOS to the above.
- Add flag to Native Assembly processor to signal the requirement of Mbed TLS regardless of having it enabled through network option.

## Motivation and Context
- Addresses nanoFramework/Home#648.
- Following nanoframework/System.Security.Cryptography#1.
- Following nanoframework/System.IO.Hashing#1.

## How Has This Been Tested?
- Unit Tests in respective libraries.

## Types of changes

- [ ] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [x] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dev Containers (changes related with Dev Containers, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist

- [x] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [x] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [x] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).

### CODE REVIEW

File: CMake/binutils.ChibiOS.cmake [hunk 1]
```diff
@@ -160,6 +160,10 @@ macro(nf_add_platform_dependencies target)
 
         endif()
 
+        if(API_nanoFramework.System.Security.Cryptography)
```
Comment (Ellerbach) [inline CMake/binutils.ChibiOS.cmake:163]: will that work as well for small devices without any network? Typically those where you want to use a modem and still benefit of crypto elements?

File: targets/ESP32/_nanoCLR/nanoFramework.System.IO.Hashing/nf_sys_io_hashing_System_IO_Hashing_Crc32.cpp [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    }
+    else
+    {
+        hash = ~esp_rom_crc32_le(~crc32, bufferData, bufferSize);
+    }
+
+#else
+
+    // this series does not have ROM support for CRC32
+    // for the time being this is not supported
```
Comment (Ellerbach) [inline targets/ESP32/_nanoCLR/nanoFramework.System.IO.Hashing/nf_sys_io_hashing_System_IO_Hashing_Crc32.cpp:57]: As some other devices may not have, I guess the next step is to build a software version for those devices, right?

**Review by Ellerbach — COMMENTED** [review]
Great work!
My question is really about boards with not HMAC compute and how to ensure we still can do it in a software way, so no device is left behind.

File: CMake/binutils.ChibiOS.cmake (hunk 1 above)
Comment (josesimoes) [reply CMake/binutils.ChibiOS.cmake:163]: Yes, this is dependent of Mbed TLS not the network stack.
As with any other depends on a build option.

File: targets/ESP32/_nanoCLR/nanoFramework.System.IO.Hashing/nf_sys_io_hashing_System_IO_Hashing_Crc32.cpp (hunk 2 above)
Comment (josesimoes) [reply targets/ESP32/_nanoCLR/nanoFramework.System.IO.Hashing/nf_sys_io_hashing_System_IO_Hashing_Crc32.cpp:57]: Yes. it's a matter of adding a lookup table for this. 
Suggest that we do than as an improvement.

Comment (Ellerbach) [reply targets/ESP32/_nanoCLR/nanoFramework.System.IO.Hashing/nf_sys_io_hashing_System_IO_Hashing_Crc32.cpp:57]: works for me. We can compute it as well (we have plenty of this in the IoT repo!)

**Review by Ellerbach — APPROVED** [review]
Thanks for this first native implementation.

**Pull request merged**